### PR TITLE
fix(hooks): eval-guard skips inline evals inside heredoc bodies (F2b)

### PR DESCRIPTION
**Pull request by jobordu**
Follow-up to #227. The eval-guard's quote scanner only tracked `'`/`"`, not heredoc delimiters — so an inline eval **mentioned inside a heredoc body** (e.g. a `git commit -F - <<'MSG' … MSG` message describing an eval bug) got rewritten and the command mangled. This kept blocking legitimate commits (hit repeatedly during this very campaign).

Added `findHeredocRanges()` (handles `<< DELIM`, `<<-DELIM`, `<< 'DELIM'`, `<< "DELIM"`); matches inside a heredoc body are skipped. Evals after a **closed** heredoc still rewrite. +3 regression tests (39/39). Found by dogfooding (the guard blocking this campaign's own commits).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed false positives in eval guard that incorrectly rewrote `node -e` commands appearing inside heredoc bodies.

### CODE REVIEW

File: hooks/nf-node-eval-guard.js [hunk 1]
```diff
@@ -79,6 +79,33 @@ function isInsideQuotes(str, index) {
...
+  const ranges = [];
+  const re = /<<(-?)\s*(['"]?)([A-Za-z_]\w*)\2/g;
+  let m;
+  while ((m = re.exec(command)) !== null) {
+    const dashLess = m[1] === '-';
+    const delim = m[3];
+    const nl = command.indexOf('\n', re.lastIndex);
+    if (nl === -1) continue; // no body on a single line
+    const bodyStart = nl + 1;
+    const closeRe = new RegExp(`^${dashLess ? '\\t*' : ''}${delim}[ \\t]*$`, 'm');
```
[the remark this answers is not in the document]
Comment (jobordu) [reply hooks/nf-node-eval-guard.js]: Good catch — fixed in 40c39409. `findHeredocRanges` now skips any `<<DELIM` that sits inside a quoted string (via the existing `isInsideQuotes`), so a quoted opener like `echo "<<'EOF'"` can no longer open a bogus body range that swallows a later real `node -e`. Added regression tests covering exactly this case.

File: hooks/nf-node-eval-guard.test.js [hunk 2]
```diff
@@ -199,6 +199,33 @@ describe('rewriteCommand quoted-substring guard', () => {
...
+    assert.ok(result.includes('console.log(1)'));
+  });
+
+  it('findHeredocRanges spans the body of quoted/dash/bare heredocs', () => {
+    assert.deepEqual(findHeredocRanges('a\n').length === 0, true);
+    const r = findHeredocRanges("x <<'E'\nbody\nE");
+    assert.equal(r.length, 1);
+    // body is "body\n" (between the opening line and the closing delimiter line)
+    assert.ok(r[0].end > r[0].start);
+  });
```
[the remark this answers is not in the document]
Comment (jobordu) [reply hooks/nf-node-eval-guard.test.js]: Added in 40c39409: `findHeredocRanges ignores a heredoc opener that sits inside quotes` (asserts no range for `echo "<<'EOF'"` and `grep '<<EOF' file`) and `still rewrites a real node -e after a quoted heredoc-opener mention`.

File: hooks/nf-node-eval-guard.js [hunk 3]
```diff
@@ -79,6 +79,37 @@ function isInsideQuotes(str, index) {
  *
  * Returns null if no rewrite needed, or the rewritten command string.
```
[the remark this answers is not in the document]
Comment (jobordu) [reply hooks/nf-node-eval-guard.js]: Fixed in 5f5cefb5 — moved `findHeredocRanges` (with its own JSDoc) above `rewriteCommand`'s JSDoc, so each doc block now sits directly on its function.

File: hooks/nf-node-eval-guard.js [hunk 4]
```diff
@@ -79,6 +79,37 @@ function isInsideQuotes(str, index) {
...
+    const rest = command.slice(bodyStart);
+    const cm = closeRe.exec(rest);
+    const bodyEnd = cm ? bodyStart + cm.index : command.length;
+    ranges.push({ start: bodyStart, end: bodyEnd });
+    re.lastIndex = bodyEnd;
+  }
+  return ranges;
+}
+
 function rewriteCommand(command) {
```
[the remark this answers is not in the document]
Comment (jobordu) [reply hooks/nf-node-eval-guard.js:117]: Fixed in 5f5cefb5 — moved `findHeredocRanges` (with its own JSDoc) above `rewriteCommand`'s JSDoc, so each doc block now sits directly on its function.

File: hooks/nf-node-eval-guard.js [hunk 5]
```diff
@@ -79,6 +79,37 @@ function isInsideQuotes(str, index) {
...
+    // `echo "<<'EOF'"`) is not a real heredoc opener — skip it so it cannot
+    // open a bogus body range that would swallow a later real `node -e`.
+    if (isInsideQuotes(command, m.index)) continue;
+    const dashLess = m[1] === '-';
+    const delim = m[3];
+    const nl = command.indexOf('\n', re.lastIndex);
+    if (nl === -1) continue; // no body on a single line
+    const bodyStart = nl + 1;
+    const closeRe = new RegExp(`^${dashLess ? '\\t*' : ''}${delim}[ \\t]*$`, 'm');
+    const rest = command.slice(bodyStart);
```
[the remark this answers is not in the document]
Comment (jobordu) [reply hooks/nf-node-eval-guard.js:102]: Intentional — the identifier charset `[A-Za-z_]\w*` (plus optional quotes) covers every heredoc delimiter that realistically wraps a body *mentioning* an inline eval: commit-message heredocs (`MSG`, `COMMIT_MSG`, `EOF`, `NF_EVAL`) are all identifiers. Broadening to arbitrary bash words risks matching non-heredoc `<<` usages and is strictly riskier for no real-world gain. Worst case of a missed exotic delimiter is a false-positive rewrite inside that body — the pre-existing behaviour — not a crash. Keeping it conservative.

File: hooks/nf-node-eval-guard.test.js [hunk 6]
```diff
@@ -199,6 +199,51 @@ describe('rewriteCommand quoted-substring guard', () => {
...
```
[the remark this answers is not in the document]
Comment (jobordu) [reply hooks/nf-node-eval-guard.test.js]: Fixed in 5f5cefb5 — split into `assert.equal(findHeredocRanges('a\n').length, 0)` under a descriptive title (`returns no ranges when there is no heredoc`).

File: CHANGELOG.md [hunk 7]
```diff
@@ -7,8 +7,8 @@ Format follows [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).
...
+- fix(hooks): `nf-node-eval-guard` no longer false-positives on an inline `node` eval that appears inside a **heredoc body** — e.g. a `git commit -F - <<'MSG' … MSG` message that merely *mentions* one. The guard's quote scanner (`isInsideQuotes`) only tracked `'`/`"`, not heredoc delimiters, so it rewrote the mention and mangled the command (this kept blocking legitimate commits whose message described an eval bug). Added `findHeredocRanges()` (handles `<< DELIM`, `<<-DELIM`, `<< 'DELIM'`, `<< "DELIM"`); matches inside a heredoc body are skipped. Real evals after a *closed* heredoc still rewrite. +3 regression tests. (Follow-up to PR #227.)
```
[the remark this answers is not in the document]
Comment (jobordu) [reply CHANGELOG.md]: Fixed in 5f5cefb5 — dropped the fragile "+3" count in favour of describing what the tests cover (heredoc-body skips, quoted-opener safety, post-heredoc rewrites).

File: hooks/nf-node-eval-guard.js [hunk 8]
```diff
@@ -73,6 +73,37 @@ function isInsideQuotes(str, index) {
...
```
[the remark this answers is not in the document]
Comment (jobordu) [reply hooks/nf-node-eval-guard.js]: Fixed in 7ca7d7ba — tightened to `^DELIM$` (and `^\t*DELIM$` for `<<-`), so a `DELIM  ` line with trailing whitespace is no longer accepted as a terminator (matching bash). Added a trailing-whitespace-terminator test asserting the body then runs to end-of-command.

File: hooks/nf-node-eval-guard.test.js [hunk 9]
```diff
@@ -199,6 +199,54 @@ describe('rewriteCommand quoted-substring guard', () => {
...
+  it('findHeredocRanges returns no ranges when there is no heredoc', () => {
+    assert.equal(findHeredocRanges('a\n').length, 0);
+  });
+
+  it('findHeredocRanges spans the body of quoted/dash/bare heredocs', () => {
+    const r = findHeredocRanges("x <<'E'\nbody\nE");
+    assert.equal(r.length, 1);
+    // body is "body\n" (between the opening line and the closing delimiter line)
+    assert.ok(r[0].end > r[0].start);
+  });
```
[the remark this answers is not in the document]
Comment (jobordu) [reply hooks/nf-node-eval-guard.test.js]: Fixed in 7ca7d7ba — the test now exercises quoted (`<<'E'`), bare (`<<E`), and dash (`<<-E` with a tab-indented terminator) forms separately, plus the trailing-whitespace case.

File: hooks/nf-node-eval-guard.js [hunk 10]
```diff
@@ -73,6 +73,40 @@ function isInsideQuotes(str, index) {
...
+    if (isInsideQuotes(command, m.index)) continue;
+    const dashLess = m[1] === '-';
+    const delim = m[3];
+    const nl = command.indexOf('\n', re.lastIndex);
+    if (nl === -1) continue; // no body on a single line
+    const bodyStart = nl + 1;
+    // Bash requires the terminator line to be EXACTLY the delimiter (only
+    // leading tabs are allowed, and only for the `<<-` form) — no trailing
+    // whitespace. Matching it exactly avoids treating `DELIM  ` as a close.
+    const closeRe = new RegExp(`^${dashLess ? '\\t*' : ''}${delim}$`, 'm');
```
[the remark this answers is not in the document]
Comment (jobordu) [reply hooks/nf-node-eval-guard.js]: Agreed — renamed to `allowLeadingTabs` in 33c9fcca (with a comment noting the `<<-` form is the one that permits a tab-indented terminator). Clearer than the double-negative `dashLess`.

File: hooks/nf-node-eval-guard.js [hunk 11]
```diff
@@ -73,6 +73,41 @@ function isInsideQuotes(str, index) {
...
+    // Bash requires the terminator line to be EXACTLY the delimiter (only
+    // leading tabs are allowed, and only for the `<<-` form) — no trailing
+    // whitespace. Matching it exactly avoids treating `DELIM  ` as a close.
+    const closeRe = new RegExp(`^${allowLeadingTabs ? '\\t*' : ''}${delim}$`, 'm');
+    const rest = command.slice(bodyStart);
+    const cm = closeRe.exec(rest);
+    const bodyEnd = cm ? bodyStart + cm.index : command.length;
+    ranges.push({ start: bodyStart, end: bodyEnd });
+    re.lastIndex = bodyEnd;
+  }
```
[the remark this answers is not in the document]
Comment (jobordu) [reply hooks/nf-node-eval-guard.js:107]: Acknowledged — this is the multiple-heredocs-on-one-line case (`cmd <<A <<B`). I'm intentionally leaving it as-is: it's vanishingly rare in practice (a single command opening two heredocs, the *second* of which contains a literal inline-eval mention), and the worst-case outcome is benign — the guard reverts to the pre-#227 behaviour of rewriting that one eval (a false-positive rewrite), not a crash, data loss, or security issue. Full stacked-heredoc body tracking would add disproportionate complexity to a fail-open safety hook for no real-world gain. The common single-heredoc forms (quoted/bare/`<<-`) are covered and correct.

File: hooks/nf-node-eval-guard.test.js [hunk 12]
```diff
@@ -199,6 +199,71 @@ describe('rewriteCommand quoted-substring guard', () => {
...
+    assert.ok(q[0].end > q[0].start);
+    // bare delimiter
+    const bare = findHeredocRanges('x <<E\nbody\nE');
+    assert.equal(bare.length, 1);
+    assert.ok(bare[0].end > bare[0].start);
+    // dash form: only leading TABS are stripped before the terminator
+    const dash = findHeredocRanges('x <<-E\nbody\n\tE');
+    assert.equal(dash.length, 1);
+    assert.ok(dash[0].end > dash[0].start);
+  });
```
[the remark this answers is not in the document]
Comment (jobordu) [reply hooks/nf-node-eval-guard.test.js:237]: Same rationale as the code thread — the `cmd <<A <<B` stacked-heredoc case is rare and fails benign (a false-positive rewrite of the mentioned eval, not a crash). I'm not adding a regression test that would pin a deliberately-unhandled edge case; the quoted/bare/`<<-` single-heredoc forms that actually occur are tested.
